### PR TITLE
Custom sounds new block option

### LINKED ISSUE #55: Custom step sounds for block creation.
**Describe the feature**

This feature would give us the option to choose a custom step sound of a block from your own imported sounds.

**How would you implement it?**

This would be a new option within the step sound drop down list.

custom

### PR DESCRIPTION
**Pull request by Goldorion**
This PR adds the possibility to create a custom sound type for blocks. This PR closes #55.

Changelog:
* Added a new subsection in the Properties page
* Move the `soundOnStep` parameter to this new subsection
* Added 5 parameters to select sounds (break, step, place, hit and fall sounds)
* Added 2 radio buttons to switch between vanilla sound types and custom sound type (like trees for biomes)

ui

Note: The Sound type class allows us to define a value for the volume and another one for the pitch, but I didn't implement them because all vanilla sound types have them set to 1.0.

Note 2: I added latest.log to the .gitignore file because otherwise, I always have to discard changes (even if nothing has changed in the file) manually.

### CODE REVIEW

**Comment by NOYBss** [comment]
Lol I was thinking about adding something like this.

File: src/main/java/net/mcreator/ui/modgui/BlockGUI.java [hunk 1]
```diff
@@ -784,6 +790,43 @@ public BlockGUI(MCreator mcreator, ModElement modElement, boolean editingMode) {
...
+				L10N.label("elementgui.soundtypes.place_sound")));
+		soundProperties.add(placeSound);
+
+		soundProperties.add(HelpUtils.wrapWithHelpButton(IHelpContext.NONE.withEntry("common/soundtypes/hit_sound"),
+				L10N.label("elementgui.soundtypes.hit_sound")));
+		soundProperties.add(hitSound);
+
+		soundProperties.add(HelpUtils.wrapWithHelpButton(IHelpContext.NONE.withEntry("common/soundtypes/fall_sound"),
+				L10N.label("elementgui.soundtypes.fall_sound")));
+		soundProperties.add(fallSound);
```
Comment (NOYBss) [inline src/main/java/net/mcreator/ui/modgui/BlockGUI.java:826]: Why did you used `IHelpContext.NONE` here for locating entries?

**Review by NOYBss — CHANGES_REQUESTED** [review]
Also why didn't you added files for the new help entries?

File: src/main/java/net/mcreator/ui/modgui/BlockGUI.java (hunk 1 above)
Comment (Goldorion) [reply src/main/java/net/mcreator/ui/modgui/BlockGUI.java:826]: In the beginning, I made the parameters in another class, so I had to use that, but after I changed them, I forgot to change that too.

**Comment by NOYBss** [comment]
Will you add the in-app tips for the new help entries in this pr, or another time?

**Comment by Goldorion** [comment]
> 
> 
> Will you add the in-app tips for the new help entries in this pr, or another time?

Maybe another time

**Review by NOYBss — COMMENTED** [review]
Ok then, looks good to me.

**Comment by Max094Reikeb** [comment]
> > Will you add the in-app tips for the new help entries in this pr, or another time?
> 
> Maybe another time

No, do it in this PR. They are related, why add them another time?

**Review by NOYBss — APPROVED** [review]
Forgot to choose "Approve" 🤦‍♂️

**Comment by Goldorion** [comment]
> 
> 
> > > Will you add the in-app tips for the new help entries in this pr, or another time?
> > 
> > 
> > Maybe another time
> 
> No, do it in this PR. They are related, why add them another time?

Ok, fine I will make them now.

**Comment by NOYBss** [comment]
> > > > Will you add the in-app tips for the new help entries in this pr, or another time?
> > > 
> > > 
> > > Maybe another time
> > 
> > 
> > No, do it in this PR. They are related, why add them another time?
> 
> Ok, fine I will make them now. Otherwise, is it good for you?

Unrelated but, I wonder if there is a limit on quoting quotes or not 🤔

**Comment by Goldorion** [comment]
> 
> 
> > > > > Will you add the in-app tips for the new help entries in this pr, or another time?
> > > > 
> > > > 
> > > > Maybe another time
> > > 
> > > 
> > > No, do it in this PR. They are related, why add them another time?
> > 
> > 
> > Ok, fine I will make them now. Otherwise, is it good for you?
> 
> Unrelated but, I wonder if there is a limit on quoting quotes or not 🤔

I don't think there is one.

File: plugins/generator-1.15.2/forge-1.15.2/templates/block.java.ftl [hunk 2]
```diff
@@ -35,6 +35,7 @@
 package ${package}.block;
 
 import net.minecraft.block.material.Material;
+import net.minecraft.util.SoundEvent;
```
Comment (Max094Reikeb) [inline plugins/generator-1.15.2/forge-1.15.2/templates/block.java.ftl:38]: Have you tested without this import?

File: .gitignore [hunk 3]
```diff
@@ -31,3 +31,4 @@ hs_err_pid*
 .DS_Store
 .AppleDouble
 .LSOverride
+logs/latest.log
```
Comment (Max094Reikeb) [inline .gitignore:34]: Could you move that to the "FILES FOR MCREATOR DEV WORKSPACE" section? Because log files aren't exclusive to MacOS...

File: plugins/generator-1.16.4/forge-1.16.4/templates/block.java.ftl [hunk 4]
```diff
@@ -35,6 +35,7 @@
 package ${package}.block;
 
 import net.minecraft.block.material.Material;
+import net.minecraft.util.SoundEvent;
```
Comment (Max094Reikeb) [inline plugins/generator-1.16.4/forge-1.16.4/templates/block.java.ftl:38]: Same here. Tested if it works without this import?

File: plugins/mcreator-localization/lang/texts.properties [hunk 5]
```diff
@@ -1051,7 +1051,6 @@ preferences.notification.checkAndNotifyForPatches.description=Uncheck this box i
   <b>WARNING: It is highly recommended to keep this enabled, or you might miss important bug fixes
 dialog.workspace.regenerate_and_build.title=Regenerate code and build
 dialog.workspace.regenerate_and_build.progress.removing_autogenerated_code=Removing autogenerated code
-dialog.workspace.regenerate_and_build.progress.moving_textures=Moving textures
```
Comment (Max094Reikeb) [inline plugins/mcreator-localization/lang/texts.properties:1054]: Why did you delete those?

File: plugins/mcreator-localization/lang/texts.properties [hunk 6]
```diff
@@ -2139,3 +2141,9 @@ dialog.tools.inject_tags.tag.items.minecraft.flowers=These items can breed bees.
 dialog.tools.inject_tags.tag.items.minecraft.small_flowers=Can be used in suspicious, fed mooshrooms, and have other properties from flowers tag
 dialog.tools.inject_tags.tag.functions.minecraft.tick=Functions tagged in this tag group will be called each world tick.
 dialog.tools.inject_tags.tag.functions.minecraft.load=Functions tagged in this tag group will be called when the world loads.
+elementgui.soundtypes.break_sound=Break sound
+elementgui.soundtypes.step_sound=Step sound
+elementgui.soundtypes.place_sound=Place sound
+elementgui.soundtypes.hit_sound=Hit sound
+elementgui.soundtypes.fall_sound=Fall sound
+elementgui.soundtypes.error_sound_type_name=Sound type needs a name
```
Comment (Max094Reikeb) [inline plugins/mcreator-localization/lang/texts.properties:2149]: Why haven't you named those like this? `elementgui.block_soundtypes.X`

**Review by Max094Reikeb — CHANGES_REQUESTED** [review]
Look at comments

File: plugins/generator-1.15.2/forge-1.15.2/templates/block.java.ftl (hunk 2 above)
Comment (Goldorion) [reply plugins/generator-1.15.2/forge-1.15.2/templates/block.java.ftl:38]: Yup, this is the good one.

File: plugins/generator-1.16.4/forge-1.16.4/templates/block.java.ftl (hunk 4 above)
Comment (Goldorion) [reply plugins/generator-1.16.4/forge-1.16.4/templates/block.java.ftl:38]: Same thing for this one

File: .gitignore (hunk 3 above)
Comment (Max094Reikeb) [reply .gitignore:34]: Oh, and I almost forgot, could you rename "FILES FOR MCREATOR DEV WORKSPACE" into "FILES FOR MCTOOLKIT DEV WORKSPACE" while you're modifying .gitignore?

File: plugins/mcreator-localization/lang/texts.properties (hunk 5 above)
Comment (Goldorion) [reply plugins/mcreator-localization/lang/texts.properties:1054]: Accident? xD

File: plugins/mcreator-localization/lang/texts.properties (hunk 6 above)
Comment (Goldorion) [reply plugins/mcreator-localization/lang/texts.properties:2149]: I will make another PR to add them to other mod elements that use the SoundType.

Comment (Max094Reikeb) [reply plugins/mcreator-localization/lang/texts.properties:2149]: then, rename those as `elementgui.common_soundtypes.X`?

File: .gitignore (hunk 3 above)
Comment (Goldorion) [reply .gitignore:34]: > 
> 
> Could you move that to the "FILES FOR MCREATOR DEV WORKSPACE" section? Because log files aren't exclusive to MacOS...

GitHub Desktop added it there automatically.

File: plugins/mcreator-localization/lang/texts.properties (hunk 5 above)
Comment (Max094Reikeb) [reply plugins/mcreator-localization/lang/texts.properties:1054]: This is not resolved yet. Plz resolve the issue before marking this conversation as resolved.

Comment (Goldorion) [reply plugins/mcreator-localization/lang/texts.properties]: I resolved the issue. I just didn't commit the changes yet.

File: plugins/mcreator-localization/lang/texts.properties (hunk 6 above)
Comment (NOYBss) [reply plugins/mcreator-localization/lang/texts.properties:2150]: instead of `_` wouldn't `.` be better?

Comment (Max094Reikeb) [reply plugins/mcreator-localization/lang/texts.properties:2150]: Maybe, `elementgui.common.soundtypes_X` (for example `elementgui.common.soundtypes_place_sound`)

Comment (Goldorion) [reply plugins/mcreator-localization/lang/texts.properties:2150]: It will be `elementgui.common.soundtypes.x`.

**Comment by Max094Reikeb** [comment]
You forgot french version for new tooltips

**Comment by Goldorion** [comment]
> 
> 
> You forgot french version for new tooltips

I didn't forgot, I won't make them.

**Comment by Max094Reikeb** [comment]
> > You forgot french version for new tooltips
> 
> I didn't forgot, I won't make them.

Why so?

**Comment by Goldorion** [comment]
Something else to add/change @Max094Reikeb ?

File: plugins/mcreator-localization/help/default/common/soundtypes/break_sound.md [hunk 7]
```diff
@@ -0,0 +1 @@
+This is the sound played when the blocks is broken.
```
Comment (Max094Reikeb) [inline plugins/mcreator-localization/help/default/common/soundtypes/break_sound.md:1]: The block in singular so it's `the block`

File: plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md [hunk 8]
```diff
@@ -0,0 +1 @@
+Ceci est le son joué lorsqu'il brisé.
```
Comment (Max094Reikeb) [inline plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md:1]: "il" should be replaced by the block. `Ceci est le son joué lorsque le bloc est brisé.`

File: plugins/mcreator-localization/help/fr_FR/common/soundtypes/fall_sound.md [hunk 9]
```diff
@@ -0,0 +1 @@
+Ceci est le son joué lorsque le bloc tombe.
```
Comment (Max094Reikeb) [inline plugins/mcreator-localization/help/fr_FR/common/soundtypes/fall_sound.md:1]: `Ce son est joué quand le bloc tombe.` would be more accurate.

File: plugins/mcreator-localization/help/fr_FR/common/soundtypes/hit_sound.md [hunk 10]
```diff
@@ -0,0 +1,2 @@
+Ceci est le son joué lorsque le joueur mine le bloc.
+Note: Ce son est une boucle qui est joué jusqu'à ce que le bloc soit détruit.
```
Comment (Max094Reikeb) [inline plugins/mcreator-localization/help/fr_FR/common/soundtypes/hit_sound.md:2]: `Note: Ce son est joué en boucle jusqu'à ce que le bloc soit détruit.`

File: plugins/mcreator-localization/help/fr_FR/common/soundtypes/step_sound.md [hunk 11]
```diff
@@ -0,0 +1 @@
+Ceci est le son joué lorsqu'une entité marche sur celui-ci.
```
Comment (Max094Reikeb) [inline plugins/mcreator-localization/help/fr_FR/common/soundtypes/step_sound.md:1]: Replace `marche sur celui-ci.` by `marche sur ce bloc.`

**Review by Max094Reikeb — CHANGES_REQUESTED** [review]
Read comments.

File: plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md (hunk 8 above)
Comment (Goldorion) [reply plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md:1]: I can keep the "il", but I will add the "est".

Comment (Max094Reikeb) [reply plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md:1]: No because what does "it" represent? The sound? Or the block?

File: plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md [hunk 12]
```diff
@@ -0,0 +1 @@
+Ceci est le son joué lorsqu'il est brisé.
```
Comment (Max094Reikeb) [inline plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md:1]: `[...] lorsque le bloc est brisé.`

**Review by Max094Reikeb — CHANGES_REQUESTED** [review]

File: plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md (hunk 12 above)
Comment (Goldorion) [reply plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md:1]: What's the reason you want `le bloc` instead of `il`?

Comment (Max094Reikeb) [reply plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md:1]: Because `le son` is in the sentence so `il` could mean `le son` but it means `le bloc` that's why.

Comment (Goldorion) [reply plugins/mcreator-localization/help/fr_FR/common/soundtypes/break_sound.md:1]: Ok, I understand. I will change it.

**Review by Max094Reikeb — APPROVED** [review]
Seems good.

**Comment by SuperMarioSega** [comment]
Cool and good
